core/src/membank_controller_io.h

```diff
@@ -0,0 +1,89 @@
+
+    bool is_address_in_range(uint16_t addr) const override {
+        return addr == controller_addr_;
+    }
+
+    uint8_t read_byte(uint16_t addr) const override {
+        if (addr == controller_addr_) {
+            // Source: https://wiki.nesdev.com/w/index.php/Standard_controller
+            // In the NES and Famicom, the top three (or five) bits are not
+            // driven, and so retain the bits of the previous byte on the bus.
+            // Usually this is the most significant byte of the address of the
+            // controller port—0x40. Certain games (such as Paperboy) rely on
+            // this behavior and require that reads from the controller ports
+            // return exactly $40 or $41 as appropriate.
+            uint8_t data = 0x40;
+            uint8_t btn = 0x00;
+
+            switch (read_cnt_++) {
+            case 0:
+                btn = controller_->get(INesController::Button::A) ? 0x01 : 0x00;
+                break;
+            case 1:
+                btn = controller_->get(INesController::Button::B) ? 0x01 : 0x00;
+                break;
+            case 2:
+                btn = controller_->get(INesController::Button::Select) ? 0x01
+                                                                       : 0x00;
+                break;
+            case 3:
+                btn = controller_->get(INesController::Button::Start) ? 0x01
+                                                                      : 0x00;
+                break;
+            case 4:
+                btn = controller_->get(INesController::Button::Up) ? 0x01
+                                                                   : 0x00;
+                break;
+            case 5:
+                btn = controller_->get(INesController::Button::Down) ? 0x01
+                                                                     : 0x00;
+                break;
+            case 6:
+                btn = controller_->get(INesController::Button::Left) ? 0x01
+                                                                     : 0x00;
+                break;
+            case 7:
+                btn = controller_->get(INesController::Button::Right) ? 0x01
+                                                                      : 0x00;
+                break;
+            }
+            data = static_cast<uint8_t>(data & static_cast<uint8_t>(0xFE)) |
+                   btn;
+            return data;
+        }
+        return 0x00;
+    }
+
+    void write_byte(uint16_t addr, uint8_t data) override {
+        if (addr == controller_addr_) {
+            if (((data_ & static_cast<uint8_t>(0x01)) == 0x01) &&
+                    ((data & static_cast<uint8_t>(0x01)) == 0x00)) {
+                // Neg flank on bit 0 should latch controller button states -
+                // here just make sure counter is reset
+                read_cnt_ = 0;
+            }
+            data_ = data;
```

Comment:
And also:
"While S (strobe) is high, the shift registers in the controllers are continuously reloaded from the button states, and reading $4016/$4017 will keep returning the current state of the first button (A). Once S goes low, this reloading will stop. Hence a 1/0 write sequence is required to get the button states, after which the buttons can be read back one at a time."
So read_byte probably needs to check the current state of S and only return button A in that case?

Reply:
Fixed, now returning state for button A as long as strobe is high.